src/main/scala/mesosphere/marathon/storage/repository/GcActor.scala

```diff
@@ -291,65 +294,68 @@ private[storage] trait ScanBehavior[K, C, S] extends StrictLogging { this: FSM[S
       podVersionsInUse.map { case (id, pods) => id -> pods.to[Set] }(collection.breakOut)
     }
 
-    def rootsInUse(): Future[Seq[StoredGroup]] = {
-      Future.sequence {
-        storedPlans.flatMap(plan =>
-          Seq(
-            groupRepository.lazyRootVersion(plan.originalVersion),
-            groupRepository.lazyRootVersion(plan.targetVersion))
-        )
-      }
-    }.map(_.flatten)
+    def rootsInUse(): Source[StoredGroup, NotUsed] = {
+      Source(storedPlans)
+        .mapConcat(plan => Seq(plan.originalVersion, plan.targetVersion))
+        .mapAsync(1)(groupRepository.lazyRootVersion)
+        .mapConcat(_.toList)
+    }
 
     def appsExceedingMaxVersions(usedApps: Set[PathId]): Future[Map[PathId, Set[OffsetDateTime]]] = {
-      Future.sequence {
-        usedApps.map { id =>
-          appRepository.versions(id).runWith(Sink.sortedSet).map(id -> _)
-        }
-      }.map(_.filter(_._2.size > maxVersions).toMap)
+      Source(usedApps)
+        .mapAsync(1)(id => appRepository.versions(id).runWith(Sink.sortedSet).map(id -> _))
```

Comment:
Hm, isn't the something like a flat map for a source of sources? I mean why do we use a sink for each app?

Reply:
We use it because we need a strict collection here.

Reply:
Hm, I found a way to do some pure streaming here.